### ISSUE TITLE
This deleted all my source files?

### ISSUE THREAD

**Issue by OmgImAlexis** [Rush/link-module-alias#3]
So I used yarn as usual to add a new dependancy and from the looks of it all my source that was symlinked via node_modules was removed.

Lucky I had just committed and could use git to retrieve the files but this looks like a huge issue.

PING: @Rush

**Comment by OmgImAlexis** [comment]
I'm testing but I believe the pre-install script is what would have prevented this. 

~Maybe there should be more than a small message in the readme about it? Personally I'd suggest having the `package` check for the `preinstall` hook and add it unless there's a flag passed.~

~Something like `--no-install`?~

**Comment by Rush** [comment]
> So I used yarn as usual to add a new dependancy and from the looks of it all my source that was symlinked via node_modules was removed.

Wow, this sounds bad. Can you show me the structure of your project? Btw. just yesterday I discovered that it's possible to link subdirectories via a `file:` version specifier, is it the way of your project?

**Comment by Rush** [comment]
> Maybe there should be more than a small message in the readme about it? Personally I'd suggest having the `package` check for the `preinstall` hook and add it unless there's a flag passed.

Can you elaborate? I am not sure I understand.

`link-module-alias` shouldn't at any circumstances remove actual source code - this needs to be fixed.

**Comment by OmgImAlexis** [comment]
Sorry I quickly wrote this issue out while trying to get my code back. I’ll update it with more details either later tonight or tomorrow.

**Comment by OmgImAlexis** [comment]
Clone [this](https://github.com/OmgImAlexis/delete-bug) repo and try the steps below. Not sure if this also happens with npm, just wanted to get this to you ASAP.

```shell
➜  delete-bug git:(master) rm -rf node_modules
➜  delete-bug git:(master) yarn
yarn install v1.12.3
warning package.json: No license field
warning No license field
[1/4] 🔍  Resolving packages...
[2/4] 🚚  Fetching packages...
[3/4] 🔗  Linking dependencies...
[4/4] 📃  Building fresh packages...

$ link-module-alias
[####################] 20/20link-module-alias: @components -> ./components
✨  Done in 0.80s.
➜  delete-bug git:(master) node main.js
TEST_COMPONENT
➜  delete-bug git:(master) yarn add fuse.js
yarn add v1.12.3
warning package.json: No license field
warning No license field
[1/4] 🔍  Resolving packages...
[2/4] 🚚  Fetching packages...
[3/4] 🔗  Linking dependencies...
[4/4] 📃  Building fresh packages...

warning No license field
success Saved 1 new dependency.
info Direct dependencies
└─ fuse.js@3.3.0
info All dependencies
└─ fuse.js@3.3.0
$ link-module-alias
link-module-alias: @components -> ./components
✨  Done in 0.82s.
➜  delete-bug git:(master) ✗ node main.js
file:///Users/xo/code/delete-bug/index.js:1
Error: Cannot find module '@components'
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:734:10)
```

**Comment by OmgImAlexis** [comment]
PING @Rush

**Comment by Rush** [comment]
@OmgImAlexis I ran your steps - very helpful! .. and it looks like it's a bug in yarn ... I'll investigate it further

**Comment by Rush** [comment]
@OmgImAlexis it's not a permanent solution but this would help avoid this particular issue:
```
   "scripts": {
     "preinstall": "which link-module-alias && link-module-alias clean || echo",
     "postinstall": "link-module-alias"
   }
```

**Comment by Rush** [comment]
@OmgImAlexis the upstream bug is #5709

I added a comment to that issue thread:
https://github.com/yarnpkg/yarn/issues/5709#issuecomment-453984390

**Comment by Rush** [comment]
@OmgImAlexis the thread seems to indicate that another workaround is not using `@` in front of the aliased package name.

**Comment by josecarneiro** [comment]
This happened to me a few times in the last week and it took me a while to understand that using yarn with `link-module-alias` was the cause of the issue. I haven't been able to reproduce it consistently, it will just happen once in a while. @Rush Is this issue only fixable on yarn's side?

**Comment by OmgImAlexis** [comment]
@josecarneiro this patch does fix the issue.

https://github.com/Rush/link-module-alias/issues/3#issuecomment-453976562

**Comment by josecarneiro** [comment]
> @josecarneiro this patch does fix the issue.
> 
> [#3 (comment)](https://github.com/Rush/link-module-alias/issues/3#issuecomment-453976562)

I added it to my `package.json` and haven't been able to reproduce the issue ever since, so it seems like a plausible solution until `yarn` solves the underlying problem.

Thank you @OmgImAlexis  and @Rush!

**Comment by josecarneiro** [comment]
Well, it did happen again after running `yarn upgrade`, even with the fix. It appears that the upgrade command does not run the preinstall hook. Just a heads up for anyone who may be encountering the same issue.

**Comment by OmgImAlexis** [comment]
This looks related. Possibly a bug?

https://github.com/yarnpkg/yarn/issues/2878

**Comment by Rush** [comment]
Is not using `@` in the package name an option for you guys?

**Comment by OmgImAlexis** [comment]
@Rush I'd prefer if we could keep using it. Any chance `@/` or `#` would work?

**Comment by josecarneiro** [comment]
I find that using ´@´ is one of the great appeals of this module. Wanting to keep consistency between my backend codebase and my webpack-built client side app is the reason I researched this topic and consequentially found ´link-module-alias´. Changing the alias is a possibility.

**Comment by shawncplus** [comment]
@Rush Having the preinstall hook do a `link-module-alias clean` fixes it for me when doing `npm install`. HOWEVER, `npm update` doesn't run the pre or post install hooks and will happily delete all of my files. I would usually dig in and help debug this but frankly after having this delete in progress work about 5 times now I'm a bit too pissed to be helpful. At the moment this project is completely useless if you can't use standard `npm install` and `npm update` commands without it deleting work.

**Comment by seanohue** [comment]
I want to add to @shawncplus 's comment that this does not seem to be yarn-related at all, since it deletes files in plain npm as well.

**Comment by jsardev** [comment]
Uh, I wasn't lucky enough to be aware of this issue before it deleted my day of work. I always forget to commit more frequently :smile:

**Comment by jsardev** [comment]
> At the moment this project is completely useless if you can't use standard `npm install` and `npm update` commands without it deleting work.

Have you tried naming your modules without the `@` prefix?

**Comment by mfaheemakhtar** [comment]
"module-alias" was not working with the ESLint, so installed this one. This module worked perfectly.

And the moment I ran "yarn remove module-alias", everything in my project directory got deleted including .git

I just have just package.json and yarn.lock in directory (and the aliased directory @utils - empty).

My local commits got lost, several hours of work. Cool!

If this module has an issue, please add a **big RED BOX** on the npm page that this can potentially wipe the  whole directory.

**Comment by Rush** [comment]
> If this module has an issue, please add a **big RED BOX** on the npm page that this can potentially wipe the whole directory.

I am very sorry you've lost your work. :-( Can you share the setup that caused everything to disappear? We can then prevent doing this particular set up or at least warning against it.

**Comment by mfaheemakhtar** [comment]
@Rush, I don't have anything because everything got wiped. So, here is the explanation - no code.

1. Directories - constants, config, utils, db, api.
2. ESLint + Prettier config files. (.eslintrc, .prettierrc).
3. jsconfig.json to make aliases work for ESLint. Had added for "module-alias" but did not work.
4. And the configuration for the module as provided in the docs. 

 "_moduleAliases": {
    "@api": "./api",
    "@constants": "./constants",
    "@config": "./config",
    "@db": "./db",
    "@root": ".",
    "@utils": "./utils"
  }

ESLint was not working with "module-alias" package, so I installed this one (yours) and ESLint worked. Then the moment I ran yarn remove module-alias, everything got deleted (not even in the Trash). 

I came to know that there is an issue with this module after it, if you could please add some warning with a link to this issue, it will be really helpful.

I'm ashamed that I had not pushed my work on GitHub. 

BTW, I'm on Linux 18.04.2.

**Comment by seanohue** [comment]
There's no need to be ashamed @mfaheemakhtar, you're in good company based on this thread. I second the idea of a huge warning on the NPM page for this module, or even unpublishing it, until the issue is resolved.

**Comment by Rush** [comment]
@seanohue @mfaheemakhtar added the warning. Let me know if you guys think it's sufficient ...

**Comment by Rush** [comment]
I think I will add a warning at runtime that discourages the use of modules starting with `@` .

**Comment by seanohue** [comment]
> I think I will add a warning at runtime that discourages the use of modules starting with `@` .

I think that's a good idea! Perhaps even throw an error if they use such an alias. I'm willing to make a PR that does this.

Thanks for being so responsive with this issue.

**Comment by Rush** [comment]
> I'm willing to make a PR that does this.

You're very welcome to give it a first shot. Thanks!

**Comment by mfaheemakhtar** [comment]
Thanks for the quick response @Rush. Really appreciate it.

Please also consider adding the warning on the npm page and change the usage example.

I read the comments above and the issues with yarn. One way to fix it could to not store the symlinks in the node_modules folder at all.

The module can create a directory for symlinks (.aliases, or anything), and add that path to the module.paths. This way (I think), the node_modules folder will be untouched and both yarn and npm should be working fine.

I'll try this module again without the "@" now, and will not add any alias for root. I guess my .git folder was removed because I had a root entry. So, even if this happens again, at least I'll have my .git folder to recover my work.

**Comment by Rush** [comment]
Hey y'all - released 1.2.0. Actions taken
- Added a runtime warning ![](https://rush.sh/0Awy2dwgBuGdOlnMh58DiXk3xjw/screenshot2.png)
- Updated examples to no longer mention `@` as a preferred prefix in aliases. I changed it to `~`. Anybody can still use `@` and they can even disable the warning. The truth is .. I tried to reproduce the issue that @mfaheemakhtar had but without success. I couldn't get my source dir removed. So maybe some people will still want to continue using `@`.

Again, I am really sorry about any data loss that might have happened. I never expected such issues to happen.

**Comment by mfaheemakhtar** [comment]
@Rush. I just tried again. 

1. GitHub    | https://github.com/mfaheemakhtar/link-module-alias-exp
2. YouTube | https://www.youtube.com/watch?v=zsDqeHXSLxI

Additional Details:
1. Node 10.15.0
2. Yarn 1.12.3
3. Ubuntu 18.04.2 64bit

I really appreciate that you took immediate action.

**Comment by jjangga0214** [comment]
# tl;dr
@Rush
~Upgrading `yarn` can resolve the bug.~ => [UPDATE : see my next comment.]

# my experience

> I tried to reproduce the issue that @mfaheemakhtar had but without success. I couldn't get my source dir removed.

Do you mean the bug does not happen on your local environment?  

Well, I've just got the accidental removal of source files today (on ubuntu 19.04). And I could repro this by the command below.

```bash
yarn install --force 
```

with this config on package.json.

```json
"scripts": {
  "postinstall": "link-module-alias",
},
"_moduleAliases": {
  "~": ".",
  "@": "src",
},
"devDependencies": {
  "link-module-alias": "^1.2.0",
},
```

So I added 

```
"preinstall": "command -v link-module-alias && link-module-alias clean || true",
```

then `yarn install --force` does not show the bug again.

However, this time the command below causes the bug.

```
yarn upgrade --force
```

**Fortunately, upgrading `yarn`(1.16.0 -> 1.17.3) does solve the problem.** No more needed to specify `preinstall` etc.

# conclusion

Merely configuring `preinstall` is not safe enough. Users should know this fact by explicit warning(on README.md?), as some of them might think they'd be "safe" if not `install`ing, but `upgrade`ing or doing other actions, which are actually potentially dangerous. In additaion, we can tell them to consider updating yarn (to +1.17.3 or certain vesion that guarantees safety).

**Comment by jjangga0214** [comment]
Oh my goodness, now the bug appears again with yarn@1.17.3. So Upgrading yarn, at least now, does not solve the problem. (As far as I know, yarn 2.0 would may fix this behavior)

For forks with the bug: I just changed `@` to `#` and it works without the bug. Personally, I feel `#` is a quite good alternative.

**Comment by Thetortuga** [comment]
Hi ! Just for helping at the fixing, the bug appeared with the + symbol instead of @.  
```
"_moduleAliases": {
    "+components": "src"
  }
```
yarn version : 1.12.3
node version: 11.14.0

(edit: but maybe I might update yarn...)

**Comment by misamisa23** [comment]
Hi, I have the same problem, this package deleted all my source code. Is there any way I can try to restore the files? 
Any help is greatly appreciated, as I have deadlines :/

**Comment by OmgImAlexis** [comment]
@misamisa23 probably not. For now I'd suggest avoiding this package.

**Comment by mfaheemakhtar** [comment]
@misamisa23, You may try a data recovery software.

**Comment by misamisa23** [comment]
> @misamisa23, You may try a data recovery software.

I'm doing just that right now. Can someone explain what exactly does this package do with the source files? Does it delete them with rm? Does it somehow delete pointers, does it move them? I need to know so I can try to debug accordingly.

**Comment by jmsansan** [comment]
Does this happen with node or is only a yarn problem? have you tried?

**Comment by mfaheemakhtar** [comment]
@misamisa23, https://github.com/yarnpkg/yarn/issues/5709#issuecomment-453984390

TLDR: This package creates symlinks in `node_modules` folder. Yarn deletes them since yarn has not created them.

**Comment by jjangga0214** [comment]
@jmsansan 
> Does this happen with node or is only a yarn problem? have you tried?

No. Node itself and node versions are irrelevant from this bug. This is caused by the behavior of yarn and npm. Commands like `yarn`, `yarn install`, `yarn upgrade`, `yarn remove`, `npm install`, `npm upgrade`, `npm uninstall` would make it happen. As far as I know, yarn 2.0 has a plan to fix this.

**Comment by jjangga0214** [comment]
@OmgImAlexis 
> @misamisa23 probably not. For now I'd suggest avoiding this package.

That could be shown as if belittling Rush's work. A "general" suggestion to avoiding a certain library should only be said when there's another "superior" alternative, or the library behaves in an "unexpected" way even with warning guide. 

Currently, link-module-alias is neither inferior or unexpected. There are no alternatives without facing other kinds of issues(e.g. working with eslint, runtime hook). Obviously this library works well with other aliases ( I personally recommend `#` instead of `@` by the way). README explicity warns data loss when using '@' at the very beginning. What's more needed?

I found this library very useful, therefore I suggest @misamisa23 to keep using it. It's sad to hear your data loss, but now you know a caveat, so you can easliy prevent an accident from now on, thus don't have to avoid using link-module-alias.

**Comment by mfaheemakhtar** [comment]
@jjangga0214, I agree. @Rush has done a great job creating this module and I also find this module great since it does not need any configurations. 

I've not tried using # yet. I'll give it a try and update my video on it.

**Comment by OmgImAlexis** [comment]
> belittling: dismissive of the importance of a person or thing.

@jjangga0214 I think you'll find I'm more worried about people's code than I am belittling @Rush. If this didn't cause people to lose work(money) then I'd have no issue using this product.

**Comment by jjangga0214** [comment]
@OmgImAlexis I don't want to make a useless debate, but let me just make it clean. 

@misamisa23 already knows he or she should use other aliases rather than '@'.

Thus, @misamisa23 would be able to use link-module-alias without such an issue. That's it.

So, it just doesn't make sense to tell him/her not to use it.

**Comment by abarriel** [comment]
Hey ! 
I tried many case, and I seems it's not `@` the issue neither the word `root` but it's when you try to alias your root folder `.`.
The alias `.` should be forbidden, but not the use of `@`.

**Comment by lwxbr** [comment]
Well, I had that problem yesterday! I lost my source files too. Luck of mine I have a repository in Bitbucket but I lost my .env that I will have to configure once again. It could be worse. I was warned but it doesn't justify the issue.

It happened when I tried to install the module `moment-timezone`. I work in Windows and that file shredding only happened when I ran `yarn add moment-timezone` in a Linux system (I don't know if it is relevant but is what it is).

**Edit:** It is clear that is not a Operating System Issue, i tried the same thing in Windows and it deleted even my .git folder.

Really, there's no way to prevent this?

We have the evidences, what we need to fix that issue for once and for all for the sake of the good operation of the npm modules? They should not be risky but safe. We need to investigate and fix!

**Comment by jsardev** [comment]
@lwxbr Please read the conversation. We know what's the root cause of this issue and it's how `yarn` and `npm` handles aliases in `node_modules`.

**Comment by lwxbr** [comment]
@sarneeh, and how you will fix it then? For me it is not clear that it is already solved. It is showing a warning that I should not use @ for alias and that's it. I'm using yarn 1.19.1 and link-module-alias 1.2.0, just update them will not solve the issue!

**Comment by jsardev** [comment]
@lwxbr As you can see this issue is not solved as it is in `Open` state. I won't fix it, because I don't have time for that and the workaround is enough for me. If you want to fix it, create a PR which solves https://github.com/yarnpkg/yarn/issues/5709 😃 Again, please read the whole conversation before asking questions which already have been answered 😉

**Comment by lwxbr** [comment]
@sarneeh, thanks for explanation! I think that issue should be closed if it is a yarn issue. But the module should create a workaround, at least one thing to prevent the deletion of the modules (that is ugly).

The questions were answered? No, they aren't in the documentation of the project. That should be there, not here hidden in a issue not resolved.

**Comment by SARFEX** [comment]
i have same problem while `npm install somenewmodule`
npm 6.13.4, node 12.14.0, Windows 10

**Comment by Rush** [comment]
> The questions were answered? No, they aren't in the documentation of the project. That should be there, not here hidden in a issue not resolved.

Agreed. Would you mind sending a PR to change the README file?

**Comment by OmgImAlexis** [comment]
A README update still doesn’t fix this issue though. This should be the default case.

This is still a danger to users that haven’t committed code. I lost a whole day worth at least a few times that’s hundred of dollars of wasted time.

**Comment by Rush** [comment]
Ok, let's put it to rest then. What is the proposed solution? For some reason people want to keep using `@` despite the warning at the top of README: https://github.com/Rush/link-module-alias#if-you-use--in-your-linked-module-names-there-is-a-risk-this-module-will-cause-data-loss

Should the warning be larger? Should it be a prompt? I don't advise using `@`

**Comment by OmgImAlexis** [comment]
I'd suggest disabling the use of the issue causing code until they enable an option that clearly states what can happen. It shouldn't be so easy for this to happen.

**Comment by EvgeniyZ** [comment]
I'm experiencing the same problem even if I'm using ```#``` sign instead of ```@```

package.json
```
"scripts": {
       ...
        "postinstall": "link-module-alias"
    },
"_moduleAliases": {
        "#pages": "./specs/pageObjects",
        "#features": "./specs/pageObjects/FeatureObjects",
        "#helpers": "./specs/helpers"
    },
"devDependencies": {
    ...
    "link-module-alias": "1.2.0",
}
```



When I'm doing ```npm install``` then files in these directories (./specs/pageObjects, ./specs/pageObjects/FeatureObjects and ./specs/helpers) are **deleted.**

**EDIT:**
I noticed that this happen only when I already had ```node_modules``` folder, so when I removed this folder and run ```npm install``` from scratch - it works fine, as expected. And after I did re-run ```npm install``` with ```node_modules``` - no data loss found. Maybe it will be helpful for someone

**Comment by hktr92** [comment]
I managed to reproduce this issue, even with [this solution](https://github.com/Rush/link-module-alias/issues/3#issuecomment-453976562).

I can't provide a testing environment, because it's a project related to workplace.

So basically this project uses multiple themes (let's give them random names: classic, acme, supernova) and shares a common JS files in the build process. 
- Themes are in `{PROJECT_ROOT}app/themes` directory, having `assets/` and `components/` subdirectories. 
- The common assets are in `{PROJECT_ROOT}core/assets/v1/js` directory
- `_moduleAlias` value is: `@sdk` -> `{PROJECT_ROOT}core/assets/v1/js` in each theme (even in `@sdk`)
- executed command `yarn remove some-useless-package` in `app/themes/supernova`
- concurrently executed command `yarn add some-cool-package` in `core/assets/v1/js`

What happened:
- the second command failed (see yarn-error.log) as it couldn't find some module in `@sdk`
- the first command was executed successfully

I think that `yarn remove some-useless-package` triggered some delete command and removed the whole alias directory (luckily I had a FTP sync in place and managed to restore from there).

Sure, it's pretty rare to have `yarn remove` command, but anyway, I have updated the `package.json` to have following lines:
```
{
    "preinstall": "which link-module-alias && link-module-alias clean || echo",
    "preuninstall": "which link-module-alias && link-module-alias clean || echo",
    "preversion": "which link-module-alias && link-module-alias clean || echo",
    "postinstall": "link-module-alias",
    "postuninstall": "link-module-alias",
    "postversion": "link-module-alias",
}
```

I think this is bloated. Is there any alternative to this, which is cleaner?

**Comment by pankajvaghela** [comment]
Maybe we can use some other directory other than node_modules for symlinks? If you want I can send PR changing the dir used for symlinks? @Rush

**Comment by OmgImAlexis** [comment]
@pankajvaghela not sure that solves the issue. Where else would they go? If not in node_modules than how exactly is the node require system meant to find them?

**Comment by pankajvaghela** [comment]
yeah my bad, figured it, after commenting. 

Maybe we can make symlinks say in folder A.  make wrapper file to export everything from there and then have that symlinked to node_modules? 
though that would work for single files or 1 level directories only

**Comment by OmgImAlexis** [comment]
The issue is intellisense and module resolution. The issue really needs to be fixed on the npm/yarn side.

**Comment by ninjz** [comment]
For any of y'all using Typescript. This article [here](https://medium.com/@rossbulat/typescript-working-with-paths-packages-and-yarn-workspaces-6fbc7087b325) gives a good solution for path aliases by configuring `tsconfig.json`

**Comment by zecka** [comment]
With typescript I solve it with [this solution](https://github.com/TypeStrong/ts-node/issues/138#issuecomment-269760415) by using [tsconfig-paths](https://www.npmjs.com/package/tsconfig-paths)

**Comment by Papooch** [comment]
So, like many others, I did not read the issue fully and thought that it only affects `yarn` and not `npm` and also lost most of the source code. I use git and commit often so it was not a huge deal, but may I suggest that the example code be changed from using `@` to use `#` or any other symbol instead? Clearly, using `@` is not safe and it should be made clearer.

**Comment by songhanpoo** [comment]
I have the same error and also lost most of the src code. I'm using `->` alternative for `@`. the bad news with an alternative symbol.

**Comment by jjangga0214** [comment]
@Rush What about letting the program print an error message and just exit when it finds `@`?

**Comment by kjbilliot** [comment]
I've had this just happen to me when using `~` as a prefix.
I was previously using `@`, but have since moved away from it after encountering the data loss issue, is there some kind of artifact left over somewhere that I should be deleting?

**Comment by lesimoes** [comment]
gosh, this package delete all my day work... This package need a serious update in readme file, please remove example with @ and put a big and clear text about this issue.

DO NOT USE @ IN ALIAS!

**Comment by Rush** [comment]
If anybody has a good idea how to update the README or code to avoid the issue please send a PR. I tried to add the warnings already

**Comment by OmgImAlexis** [comment]
Why not remove the ability to use it? That might be the best option here.
